### ISSUE TITLE
How to infer on a Torch Tensor on GPU and return same Torch Tensor on GPU

### ISSUE THREAD

**Issue by harshitsankhla** [NVIDIA/TensorRT#2506]
I have 2 TRTEngine converted models, lets say encoder and decoder. I would like to infer from both models in tandem. The input to the encoder is TorchTensor which is already on the GPU. I would like to pass this to the encoder TRT engine and then its output to the decoder.

Flow can be understood as - 
 inp = torch.ones((1, 64, 200, 200)).cuda()
out1 = encoder.infer(x)
some_non_deep_operations(out1)
out2 = decoder.infer(out1)

Right now the way I am inferring from each engine is - 

```
def infer(self, img):
		# np.copyto(self.host_inputs[0], img.ravel())
		self.host_inputs[0] = img.data.cpu().numpy().astype(np.float32)

		if self.cuda_ctx:
			self.cuda_ctx.push()
		
		cuda.memcpy_htod_async(self.cuda_inputs[0], self.host_inputs[0], self.stream)

		self.context.execute_async(
			batch_size=1,
			bindings=self.bindings,
			stream_handle=self.stream.handle)
		
		cuda.memcpy_dtoh_async(self.host_outputs[0], self.cuda_outputs[0], self.stream)
		
		self.stream.synchronize()
		
		if self.cuda_ctx:
			# print("###Cleaned context")
			self.cuda_ctx.pop()
		
		# print("###Inferred")	
		# print(type(self.host_outputs), len(self.host_outputs))

		return self.host_outputs[0]
```

Right now, in its current state, this method can input a torch tensor, convert it to a numpy array, move it to the CPU, then move it again to the GPU(htod_async), then infer, then move the result back to CPU(dtoh_async) and return the output as a numpy array on the CPU, which I have to again convert to a Torch Tensor and move to the GPU. 

Following are my questions/confusions, and would appreciate any insights on them. 

1. How to pass a torch tensor on the GPU directly to infer on the engine? (From #2456 it seems I can set the bindings as img.data_ptr()), however the output still has to be moved back to a output buffer and it returns as numpy array. 
2. Can I not even have to call dtoh_async and do an inplace transform on the input and continue using it for the second model.
3. Can the device_output be directly returned as torch tensor somehow? 

Ideally I would want to pass the TorchTensor which has been loaded on the GPU, infer on it and continue using it at the same memory address, just reshaping it whenever needed.

Any help will be appreciated. 
@pranavm-nvidia

**Comment by zerollzeng** [comment]
@pranavm-nvidia ^ ^ I don't know much about this question, can you help here?

**Comment by pranavm-nvidia** [comment]
You could do e.g.
```py
context.execute_async_v2(bindings=[img.data_ptr(), img.data_ptr()], ...)
```
to use `img` for both the input and output.

**Comment by harshitsankhla** [comment]
> You could do e.g.
> 
> ```python
> context.execute_async_v2(bindings=[img.data_ptr(), img.data_ptr()], ...)
> ```
> 
> to use `img` for both the input and output.

I had tried this, but it does not work. In fact, it does not even perform the operation.

**Comment by nvpohanh** [comment]
> I had tried this, but it does not work. In fact, it does not even perform the operation.

Could you show your code about how to did this?

Some things to check first:

- What "stream" did you use? Did you use the same stream as PyTorch uses? 
- Try add a stream synchronization after execute_async_v2 and see what the output values are?

**Comment by harshitsankhla** [comment]
Pasting the code that summaries the flow and operations

```
with open("TRT_files/bevencode.trt", 'rb') as f, trt.Runtime(trt.Logger(trt.Logger.WARNING)) as runtime:
			self.bev_engine = runtime.deserialize_cuda_engine(f.read())     # decoder
with open("TRT_files/camencode.trt", 'rb') as f, trt.Runtime(trt.Logger(trt.Logger.WARNING)) as runtime:
			self.cam_engine = runtime.deserialize_cuda_engine(f.read())    # encoder
	
self.bev_outputs = torch.ones((1, 160000), device="cuda:0")   # tensor to store the output of decoder
self.cam_outputs = torch.ones((1, 2770944), device="cuda:0") # tensor to store the output of encoder
		
self.stream = torch.cuda.Stream(device=self.device)  # device stream

# encoding part
bindings = [int(images.data_ptr()), int(self.cam_outputs.data_ptr())]
self.cam_context = self.cam_engine.create_execution_context()
self.cam_context.execute_async_v2(bindings, stream_handle=self.stream.cuda_stream)
self.stream.synchronize()
images = self.cam_outputs.reshape((1, 6, 64, 41, 8, 22))

# non-deep operations
x = geometric_operation(images)

# decoding part
bindings = [int(x.data_ptr()), int(self.bev_outputs.data_ptr())]
self.bev_context = self.bev_engine.create_execution_context()
self.bev_context.execute_async_v2(bindings, stream_handle=self.stream.cuda_stream)
self.stream.synchronize()
final_out = self.bev_outputs.reshape((1, 4, 200, 200))
```

the output is completely blank

**Comment by sandeepkumar-skb** [comment]
Try doing the following:
```
images = self.cam_outputs.reshape((1, 6, 64, 41, 8, 22)).clone()
```

```

**Comment by harshitsankhla** [comment]
> Try doing the following:
> 
> ```
> images = self.cam_outputs.reshape((1, 6, 64, 41, 8, 22)).clone()
> ```
> 
> ```
> final_out = self.bev_outputs.reshape((1, 4, 200, 200)).clone()
> ```

This also does not work. By the way, I checked, the output tensor has NAN values in some indices.

**Comment by nvpohanh** [comment]
Could you also check the binding dtype of the engine?

https://docs.nvidia.com/deeplearning/tensorrt/api/python_api/infer/Core/Engine.html#tensorrt.ICudaEngine.get_tensor_dtype

I am worried that your engines were built with FP16 inputs/outputs, but `torch.ones()` actually gives you a FP32 tensor.

**Comment by harshitsankhla** [comment]
> Could you also check the binding dtype of the engine?
> 
> https://docs.nvidia.com/deeplearning/tensorrt/api/python_api/infer/Core/Engine.html#tensorrt.ICudaEngine.get_tensor_dtype
> 
> I am worried that your engines were built with FP16 inputs/outputs, but `torch.ones()` actually gives you a FP32 tensor.

So I tried this - 

`import tensorrt as trt
with open("TRT_files/bevencode.trt", 'rb') as f, trt.Runtime(trt.Logger(trt.Logger.WARNING)) as runtime:
	bev_engine = runtime.deserialize_cuda_engine(f.read())
for binding in bev_engine:
    print(bev_engine.get_binding_dtype(binding))
    print(trt.nptype(bev_engine.get_binding_dtype(binding)))`

The output is -

`DataType.FLOAT
<class 'numpy.float32'>
DataType.FLOAT
<class 'numpy.float32'>`

Based on this I changed the dtype for the torch.ones() by setting dtype=torch.float32

Still blank output. Also observed, the output values are all NANs now.

**Comment by nvpohanh** [comment]
Hmm so it seems that the bindings are indeed in FP32.

Could you share the whole repro steps, including the input data you used? We may need to debug this...

**Comment by harshitsankhla** [comment]
For input data - 

```
from PIL import Image
imgs = []
        for imgname in img_names:
            img = Image.open(os.path.join(dataroot, f'{imgname}.jpg'))
            img = torchvision.transforms.ToTensor()(img)
            img.to(device)
            img = torchvision.transforms.Resize((resize_dims[1], resize_dims[0]))(img)
            img = torchvision.transforms.functional.crop(img, crop_h, crop_w, fH, fW)
            img = torchvision.transforms.Normalize(mean=[0.485, 0.456, 0.406], std=[0.229, 0.224, 0.225])(img)
            imgs.append(img)
        images = torch.stack(imgs).type(torch.DoubleTensor).unsqueeze(0).to(device)  # this will have 6 images
```

After this comes the TRT inference part, for which the code is [shared](https://github.com/NVIDIA/TensorRT/issues/2506#issuecomment-1334909705) in the comments up.

That's pretty much it. The TRT engine conversion happens by first converting to ONNX then to TRT, and that should not be the problem cause with the original way inference by GPU-CPU-GPU transfer is working fine.

**Comment by harshitsankhla** [comment]
> > You could do e.g.
> > ```python
> > context.execute_async_v2(bindings=[img.data_ptr(), img.data_ptr()], ...)
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > to use `img` for both the input and output.
> 
> I had tried this, but it does not work. In fact, it does not even perform the operation.

@pranavm-nvidia another reason this will not work is both the input tensor and output tensor have different dimensions which cannot be recovered just by reshaping as the final output of the network has a reduced volume size due to pooling operations

**Comment by nvpohanh** [comment]
I received a small repro from @harshitsankhla . Now debugging it...

**Comment by harshitsankhla** [comment]
Created a small gist [here](https://gist.github.com/harshitsankhla/80012a96b47e1a34e191ed66dea755b1), for any souls getting stuck in the same challenge

**Comment by nvpohanh** [comment]
TensorRT doesn't support different input sizes within a batch. You will need to either run each input sample one-by-one or pad the smaller ones to the maximum size withing the batch.

**Comment by nvpohanh** [comment]
> How can I run each sample individually if I have the output of the encoder (also the input for the decoder) like this?

Just build a TRT engine with BS1 and call `execute` twice, each time with one input tensor? Would be great if you can show the code that you already have.

> In the case of the padding method, does the architecture of the encoder and decoder also have to be modified?

Yes, I think so

**Comment by zmtttt** [comment]
when have muti  inputs and outputs? how to deal it? ex 5 inputs and 2ouputs

**Comment by finlay-hudson** [comment]
For a method updated for tensorrt 10, and accepting multiple inputs and outputs 

```
from pathlib import Path

from cuda import cudart
import torch
import tensorrt as trt


class CudaStream:
    def __enter__(self):
        err, self._stream = cudart.cudaStreamCreate()
        assert err == 0

        return self._stream

    def __exit__(self, type, value, traceback):
        cudart.cudaStreamDestroy(self._stream)


class TRTRunner:
    def __init__(self, engine_fp: Path | str, out_shapes: dict[str, tuple[torch.dtype, tuple]],
                 device=torch.device("cuda:0")) -> None:
        """
        :param engine_fp: path pointing to tensorrt engine file
        :param out_shapes: dict of form {"output_name": (output_dtype, output_shape)}
        :param device: torch.device
        """

        self.device = device
        self.runtime = trt.Runtime(trt.Logger(trt.Logger.INFO))
        with open(engine_fp, "rb") as f:
            self.engine = self.runtime.deserialize_cuda_engine(f.read())
        self.out_shapes = out_shapes

        self.ctx = self.engine.create_execution_context()
        torch.cuda.synchronize()

    def run(self, inputs: dict[str, torch.Tensor | list[tuple[torch.Tensor, torch.Tensor]]]) -> torch.Tensor | tuple:
        """
        :param inputs: dict of form {"input_name": torch.Tensor | list[tuple[torch.Tensor, torch.Tensor]]}
        :return: torch.Tensor | tuple
        """
        ctx = self.ctx

        def map_input_tensor(name, v):
            ctx.set_input_shape(name, v.shape)
            ctx.set_tensor_address(name, v.data_ptr())

        def walk_input(name, v):
            if isinstance(v, (list, tuple)):
                for i, v_1 in enumerate(v):
                    walk_input(f"{name}.{i}", v_1)
            else:
                map_input_tensor(name, v)

        for k, inp in inputs.items():
            walk_input(k, inp)

        outputs = []
        for k, (dtype, shape) in self.out_shapes.items():
            o = torch.zeros(shape, dtype=dtype, device=self.device)
            outputs.append(o)
            ctx.set_tensor_address(k, o.data_ptr())

        assert len(ctx.infer_shapes()) == 0
        with CudaStream() as stream:
            ctx.execute_async_v3(stream)
            cudart.cudaStreamSynchronize(stream)

        if len(outputs) == 1:
            return outputs[0]
        else:
            return tuple(outputs)
```
